backend/src/plonky3/stark.rs

```diff
@@ -205,22 +205,35 @@ where
         witness: &[(String, Vec<T>)],
         witgen_callback: WitgenCallback<T>,
     ) -> Result<Vec<u8>, String> {
-        // here we need to clone the witness because the callback will modify it
-        let witness = &mut witness.to_vec();
+        let mut witness_by_machine = self
+            .split
+            .iter()
+            .map(|(machine, (pil, _))| {
+                (
+                    machine.clone(),
+                    machine_witness_columns(witness, pil, machine),
```

Comment:
There's also cloning inside this method, I wonder if we could get rid of it but it seems like it would require moving to machine-level witnesses everywhere?